### ISSUE TITLE
Research: Explore the possibility to expose a callback function when doPrompting is false

### ISSUE THREAD

**Issue by zFernand0** [zowe/imperative#422]
The `addPropsOrPrompt` works great for CLI as the client of this functionality.
There is an opportunity to improve the SDK further so that people can implement their own mechanism for asking end users for information *at the right time*. For this purpose, we could evaluate what would it take to provide a callback mechanism instead of skipping the gathering of required properties.

https://github.com/zowe/imperative/blob/c9de181194933c2ba82e8d85513a26aa3a87b445/packages/rest/src/session/ConnectionPropsForSessCfg.ts#L123

**Comment by gejohnston** [comment]
Here are some implementation thoughts on this subject:

- The function that we would modify is: 

    imperative\packages\rest\src\session\ConnectionPropsForSessCfg.ts - 
    ConnectionPropsForSessCfg.addPropsOrPrompt()

- In the CLI, addPropsOrPrompt is called from a command handler like this:
```
const sessCfg: ISession =  {
    rejectUnauthorized: commandParameters.arguments.rejectUnauthorized,
    basePath: commandParameters.arguments.basePath
};
const connectableSessCfg = await ConnectionPropsForSessCfg.addPropsOrPrompt<ISession>(
     sessCfg, commandParameters.arguments
);
mySession = new Session(connectableSessCfg);
```

- This is the signature of the addPropsOrPrompt function:
```
public static async addPropsOrPrompt<T>(
    initialSessCfg: T,
    cmdArgs: ICommandArguments,
    options: IOptionsForAddConnProps = {}
): Promise<T>
```

- This is the definition for IOptionsForAddConnProps:
```
export interface IOptionsForAddConnProps {
    requestToken?: boolean;
    doPrompting?: boolean;
    defaultTokenType?: SessConstants.TOKEN_TYPE_CHOICES;
}
```

- When not supplied, the default value for doPrompting is true, so the CLI currently prompts without having to set this option.
    - With the new option proposed below, the doPrompting option may become unnecessary, but I think that we would keep the doPrompting option, so that we do not introduce a breaking change.

- We could add a new option property named getValuesCallback. It would be a function that does its own prompting.
    - The getValuesCallback function must accept one parameter named 'neededProperties'. It would be a map like the following:

    ```
        {
            user: null,     <--  example
            neededProp2: null,
            neededProp3: null
        }
    ```

    - The addPropsOrPrompt function would form the neededProperties object with the property names for which we must prompt, and then call getValuesCallback(neededProperties).

    - The getValuesCallback function will replace the nulls with the values that it gets from prompting.

- When Zowe Explorer is creating a session, it would call ConnectionPropsForSessCfg.addPropsOrPrompt() and set the following options:

        options.doPrompting = false
        options.getValuesCallback = your own function

- The CLI would *NOT* supply a callback.  addPropsOrPrompt will just do its current command-line prompting if  options.doPrompting is true and options.getValuesCallback is not defined.

- The Imperative (and CLI) programming changes should be confined to addPropsOrPrompt()
    - Test for the existence of the new getValuesCallback option
        - Form the neededProperties object
        - Call getValuesCallback(neededProperties)
        - Place the neededProperties values into the session configuration

Maybe others will eventually find additional actions that must be implemented to make this work, but I think that this is a starting point for our thinking on this subject.

**Comment by jellypuno** [comment]
@zFernand0 @gejohnston @katelynienaber   I'm working on this and I am still having a hard time understanding it. I apologize for being slow. I would really appreciate your help in clarifying some things for me.

I have a few questions:

1) What parameters should I put in neededProperties? all Profile values? which profile? or should it be generic? If it is generic, how many properties should I put?
2) What does this mean "Place the neededProperties values into the session configuration"?

**Comment by katelynienaber** [comment]
I think neededProperties should contain the things you want to prompt for. So ex. if my profile contains:
```
host: ""
port: 0
user: ""
password: ""
basePath: api/v1
rejectUnauthorized: false
```

Then I need to pass in neededProperties object `{ host, port, user, password }` because those are the properties that are missing.

And then the values we get from the user in getValuesCallback should be assigned to the object.

I am doing this all right now like so (using an array to store neededProperties instead of an object):
```
            const schemaArray = [];

            // Select for prompting only fields which are not defined
            if (!serviceProfile.user) { schemaArray.push("user"); }
            if (!serviceProfile.password) { schemaArray.push("password"); }
            if (!serviceProfile.host) {
                schemaArray.push("host");
                if (!serviceProfile.port) { schemaArray.push("port"); }
                if (!serviceProfile.basePath) { schemaArray.push("basePath"); }
            }
            const newDetails = await this.collectProfileDetails(profileName, null, schemaArray);
            for (const detail of schemaArray) { serviceProfile[detail] = newDetails[detail]; }
```

The values each profile needs are sometimes different. Currently in Explorer you can get the values available on a certain profile by calling `getSchema(profileType)` in the Profiles class.

**Comment by jellypuno** [comment]
@katelynienaber ok. perfect! thank you. I'll see what I can do.

**Comment by zFernand0** [comment]
@gejohnston, 
I do like the idea of completely separating the `stdin` prompting logic from that of the application (vscode extension, web app, ...) by doing `options.doPrompting = false`.
However, I'd like to propose that imperative asks the application for each individual property that it still requires. Here is how...
In the following places:
https://github.com/zowe/imperative/blob/67e29b1502f1afa2e1887b9871370372f20520c2/packages/rest/src/session/ConnectionPropsForSessCfg.ts#L122-L125
https://github.com/zowe/imperative/blob/67e29b1502f1afa2e1887b9871370372f20520c2/packages/rest/src/session/ConnectionPropsForSessCfg.ts#L174-L179

We could embed the callback, so that we don't do the stdin prompting.
This allows the application developers to continue to `doPrompting = true` and get called each time there is a property missing. Hopefully this will help in the future whenever we prompt for credentials if the token expires.
```javascript
                let answer = "";
                if (cmdArgs.getValueCallback != null) {
                  answer = await cmdArgs.getValueCallback("profile.key");
                } else {
                  while (answer === "") {
                      answer = await CliUtils.promptWithTimeout(
                          "Enter the host name of your service: "
                      );
                      if (answer === null) {
                          throw new ImperativeError({msg: "Timed out waiting for host name."});
                      }
                  }
                }
                finalSessCfg.hostname = answer;
```

`"profile.key"` is one of the following `host | port | user | password`

**Comment by gejohnston** [comment]
@zFernand0
It is reasonable to use the existence of getValueCallback as the trigger to call the callback function.
The only concern about calling the callback for each individual property would be that a GUI app might want to display a single dialog in which the user fills in all of the needed values at once, rather than pop-up a question for each property from individual calls to the callback.

**Comment by zFernand0** [comment]
That makes sense. I guess in that case we will have to duplicate some of the logic to check which parameters are missing (similar to what Katelyn suggested. 🤔
I do see benefit in both though and having a `doPrompting = false` will also bypass all the individual calls to the callback that I'm proposing 😋 
@jellypuno, Would it be ok to do both? And only replace the individual prompts if doPrompting happens to be false?

(personal opinion) I feel like a quick MVP would be the individual calls (piggybacking off of Imperative's logic already implemented, hence keeping it centralized)?

**Comment by jellypuno** [comment]
@zFernand0 So as we discussed yesterday, Base Profiles are not automatically loaded when we call the addProps function. It would be great if imperative can load it for us. But for now what I am going to do is:
1. read the base profle
2. Pass the Service and Base Profile to addProps function
3. Call my function to validate the missing fields.

Am I correct?

**Comment by tjohnsonBCM** [comment]
@zFernand0 Can this issue be closed since it has been addressed by #433 and #436?

**Comment by zFernand0** [comment]
> @zFernand0 Can this issue be closed since it has been addressed by #433 and #436?

I believe so. :yum:
